File: base/src/main/java/com/yxr/base/util/ToastUtil.java
package com.yxr.base.util;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.os.Message;
import android.text.TextUtils;
import android.widget.Toast;

import com.yxr.base.BaseApplication;

import java.lang.reflect.Field;

/**
 * @author ciba
 * @description 安全的Toast工具类
 * @date 2020/09/17
 */
public class ToastUtil {
    public static void show(String message) {
        show(BaseApplication.context, message);
    }

    public static void show(Context context, String message) {
        if (context != null && !TextUtils.isEmpty(message)) {
            context = context.getApplicationContext();
            try {
                if (Build.VERSION.SDK_INT == Build.VERSION_CODES.N_MR1) {
                    SafeToastUtil.show(context, message);
                } else {
                    Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @SuppressLint("SoonBlockedPrivateApi")
    public static class SafeToastUtil {
        private static Field sField_TN;
        private static Field sField_TN_Handler;

        static {
            try {
                sField_TN = Toast.class.getDeclaredField("mTN");
                sField_TN.setAccessible(true);
                sField_TN_Handler = sField_TN.getType().getDeclaredField("mHandler");
                sField_TN_Handler.setAccessible(true);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        public static void show(Context context, String message) {
            Toast toast = Toast.makeText(context, message, Toast.LENGTH_SHORT);
            hook(toast);
            toast.show();
        }

        private static void hook(Toast toast) {
            try {
                Object tn = sField_TN.get(toast);
                Handler preHandler = (Handler) sField_TN_Handler.get(tn);
                sField_TN_Handler.set(tn, new SafelyHandlerWrapper(preHandler));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        private static class SafelyHandlerWrapper extends Handler {

            private Handler impl;

            SafelyHandlerWrapper(Handler impl) {
                this.impl = impl;
            }

            @Override
            public void dispatchMessage(Message msg) {
                try {
                    super.dispatchMessage(msg);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void handleMessage(Message msg) {
                //需要委托给原Handler执行
                if (impl != null) {
                    impl.handleMessage(msg);
                }
            }
        }
    }
}

File: base/src/main/java/com/yxr/base/util/ToastUtil.kt
package com.yxr.base.util

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.text.TextUtils
import android.widget.Toast
import com.yxr.base.BaseApplication
import java.lang.reflect.Field

class ToastUtil {
    companion object {
        @JvmStatic
        fun show(message: String?) {
            if (!TextUtils.isEmpty(message)) {
                try {
                    if (Build.VERSION.SDK_INT == Build.VERSION_CODES.N_MR1) {
                        SafeToastUtil.show(BaseApplication.context, message)
                    } else {
                        Toast.makeText(BaseApplication.context, message, Toast.LENGTH_SHORT).show()
                    }
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
            }
        }

        @SuppressLint("SoonBlockedPrivateApi")
        object SafeToastUtil {
            private var sField_TN: Field? = null
            private var sField_TN_Handler: Field? = null

            init {
                try {
                    sField_TN = Toast::class.java.getDeclaredField("mTN")
                    val accessible = sField_TN?.setAccessible(true)
                    sField_TN_Handler = sField_TN?.type?.getDeclaredField("mHandler")
                    sField_TN_Handler?.isAccessible = true
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
            }

            fun show(context: Context?, message: String?) {
                val toast = Toast.makeText(context, message, Toast.LENGTH_SHORT)
                hook(toast)
                toast.show()
            }

            private fun hook(toast: Toast) {
                try {
                    val tn = sField_TN!![toast]
                    val preHandler = sField_TN_Handler!![tn] as Handler
                    sField_TN_Handler!![tn] = SafelyHandlerWrapper(preHandler)
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
            }

            private class SafelyHandlerWrapper internal constructor(private val impl: Handler?) :
                Handler(Looper.getMainLooper()) {
                override fun dispatchMessage(msg: Message) {
                    try {
                        super.dispatchMessage(msg)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }

                override fun handleMessage(msg: Message) {
                    //需要委托给原Handler执行
                    impl?.handleMessage(msg)
                }
            }
        }
    }
}